src/FFmpegWriter.cpp

```diff
@@ -1361,23 +1352,51 @@ void FFmpegWriter::open_video(AVFormatContext *oc, AVStream *st) {
 
 #if IS_FFMPEG_3_2
 	if (hw_en_on && hw_en_supported) {
-		video_codec->max_b_frames = 0;        // At least this GPU doesn't support b-frames
-		video_codec->pix_fmt   = hw_en_av_pix_fmt;
-		video_codec->profile = FF_PROFILE_H264_BASELINE | FF_PROFILE_H264_CONSTRAINED;
-		av_opt_set(video_codec->priv_data,"preset","slow",0);
-		av_opt_set(video_codec->priv_data,"tune","zerolatency",0);
-		av_opt_set(video_codec->priv_data, "vprofile", "baseline", AV_OPT_SEARCH_CHILDREN);
+		switch (video_codec->codec_id) {
+			case AV_CODEC_ID_H264:
+				video_codec->max_b_frames = 0;        // At least this GPU doesn't support b-frames
+				video_codec->pix_fmt   = hw_en_av_pix_fmt;
+				video_codec->profile = FF_PROFILE_H264_BASELINE | FF_PROFILE_H264_CONSTRAINED;
+				av_opt_set(video_codec->priv_data, "preset", "slow", 0);
+				av_opt_set(video_codec->priv_data, "tune", "zerolatency", 0);
+				av_opt_set(video_codec->priv_data, "vprofile", "baseline", AV_OPT_SEARCH_CHILDREN);
+				break;
+			case AV_CODEC_ID_HEVC:
+				// must be zero (at least on AMDGPU on Linux 8/2019)
+				video_codec->bit_rate = 0;
+
+				//TODO: this can be made configurable (25 default)
+				//av_opt_set(video_codec->priv_data,"qp","27",0);
+
+				// FYI other hevc_vaapi ffmpeg defaults are this
+				/*
+					video_codec->max_b_frames = 2;
+					video_codec->gop_size = 120;
+					video_codec->i_quant_factor = 1;
+					video_codec->i_quant_offset = 0;
+					video_codec->b_quant_factor = 6.0/5;
+					video_codec->qmin = -1;
+					video_codec->qmax = -1;
+				*/
+				break;
+			default:
+				ZmqLogger::Instance()->AppendDebugMethod("No codec-specific options defined for this codec. HW encoding may fail",
+					"codec_id", video_codec->codec_id);
+				break;
+		}
+
 		// set hw_frames_ctx for encoder's AVCodecContext
 		int err;
 		if ((err = set_hwframe_ctx(video_codec, hw_device_ctx, info.width, info.height)) < 0) {
-				fprintf(stderr, "Failed to set hwframe context.\n");
+				ZmqLogger::Instance()->AppendDebugMethod("FFmpegWriter::open_video ERROR faled to set hwframe context",
+					"width", info.width, "height", info.height);
```

Comment:
This could include more context, even — something like
"FFmpegWriter::open_video (set_hwframe_ctx) ERROR faled to set hwframe context"
(Just to make the call searchable within the code. But it's fine as-is, too.)

I was going to suggest adding an `av_err2str(err)` argument to the logging call, but the log redirection (your other PR, hopefully upcoming? :grin: ) should take care of passing along any error message that results from a failed call, so no sense in creating duplicate messages.

Reply:
it is actually a good idea - these errors are "higher-level" enum-like, useful to give some generic error overview. Full ffmpeg is verbose and I am not sure if it should be enabled every time a OpenShot debug log is enabled. Let's talk about it in the separate error PR but I will include av_err2str here just in case.

Reply:
> Full ffmpeg is verbose and I am not sure if it should be enabled every time a OpenShot debug log is enabled. 

At the moment, the big problem is that debug libopenshot is even _more_ verbose — and as things stand it's very much all-or-nothing, either you have debugging turned on and you get _flooded_ with messages for every frame encoded, or you don't even get ERROR messages when something goes wrong on the ffmpeg end. We need to strike a better balance, or implement some sort of multi-level message passing (not sure if it should be level-setting or multiple distinct ZMQ streams, but _something_)... not there yet, though.

> Let's talk about it in the separate error PR but I will include av_err2str here just in case.

Sounds good, thanks!